core/trino-main/src/main/java/io/trino/execution/scheduler/FullNodeCapableNodeAllocatorService.java

```diff
@@ -0,0 +1,616 @@
+            SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+            future.set(node);
+            if (future.isCancelled()) {
+                releaseFullNode(node, pendingAcquire.getQueryId());
+            }
+        });
+
+        failures.forEach((pendingAcquire, failure) -> {
+            SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+            future.setException(failure);
+        });
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        stopped.set(true);
+        executor.shutdownNow();
+    }
+
+    public synchronized Optional<InternalNode> tryAcquire(NodeRequirements requirements, Candidates candidates, QueryId queryId)
+    {
+        if (isFullNode(requirements)) { // todo
+            return tryAcquireFullNode(candidates, queryId);
+        }
+        return tryAcquireSharedNode(candidates, requirements.getMemory().toBytes());
+    }
+
+    @VisibleForTesting
+    synchronized Set<InternalNode> getPendingFullNodes()
+    {
+        return ImmutableSet.copyOf(fullNodePendingAcquires.keySet());
+    }
+
+    private synchronized Optional<InternalNode> tryAcquireFullNode(Candidates candidates, QueryId queryId)
+    {
+        Collection<InternalNode> queryFullNodes = fullNodesByQueryId.get(queryId);
+
+        if (fullNodesCountExceeded(queryFullNodes.size(), candidates.getAllNodesCount())) {
+            return Optional.empty();
+        }
+
+        // select nodes which are not used nor waited for
+        Optional<InternalNode> selectedNode = candidates.getCandidates().stream()
+                .filter(node -> sharedAllocatedMemory.getOrDefault(node, 0L) == 0)
+                .filter(node -> !allocatedFullNodes.contains(node))
+                .filter(node -> !fullNodePendingAcquires.containsKey(node))
+                .findFirst();
+
+        selectedNode.ifPresent(node -> {
+            allocatedFullNodes.add(node);
+            fullNodesByQueryId.put(queryId, node);
+        });
+
+        return selectedNode;
+    }
+
+    private boolean fullNodesCountExceeded(int currentCount, int candidatesCount)
+    {
+        long threshold = Integer.min(maxAbsoluteFullNodesPerQuery, (int) (candidatesCount * maxFractionFullNodesPerQuery));
+        return currentCount >= threshold;
+    }
+
+    private synchronized Optional<InternalNode> tryAcquireSharedNode(Candidates candidates, long memoryLease)
+    {
+        Optional<InternalNode> selectedNode = candidates.getCandidates().stream()
+                .filter(node -> node.getMemoryPoolSize().toBytes() - sharedAllocatedMemory.getOrDefault(node, 0L) >= memoryLease) // not enough memory on the node
+                .filter(node -> !allocatedFullNodes.contains(node)) // node is used exclusively
+                .filter(node -> !fullNodePendingAcquires.containsKey(node)) // flushing node to get exclusive use
+                .min(comparing(node -> sharedAllocatedMemory.getOrDefault(node, 0L)));
```

Comment:
QQ: is worst-fit the default bin-packing algorithm in Trino and has proven to work better than best-fit in practice?

Reply:
Good question. I do not know if there were any experiments really. I see (checked now) that for standard exuetion mode (for interactive queries without retries) we use similar approach (yet based on split weight/count not memory):
https://github.com/trinodb/trino/blob/b1e66fc4137a6aba57ec8b76870c258f1cd499f7/core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java#L148-L150